docs: Promote `when-then-otherwise` in User Guide

dangotbanned opened the pull request:
Will close #3500

# Status
All outstanding work should be scoped to `parameters.rst`.
*Corrections/suggestions elsewhere are still welcome*

The [linked version](https://github.com/dangotbanned/altair/blob/390d59935bf1cf81ce4988fdc6099b9d9a50b790/doc/user_guide/interactions/parameters.rst) represents **Draft 1**.
- Kept things mostly as they were for `alt.condition`
- Hoping to get feedback on
	- **How much** of [alt.when](https://altair-viz.github.io/user_guide/generated/api/altair.when.html#altair.when) feels **missing**?
	- If something is missing, can a link to an example improve the read?
- Potentially missing
	- Chaining, reuse e.g. https://github.com/vega/altair/pull/3544/files#r1746885701
	- Simply repeating these in order https://github.com/vega/altair/pull/3492#pullrequestreview-2205990789

# Tasks
- [x] Use `alt.when` in [waterfall chart](https://github.com/vega/altair/issues/3500#issuecomment-2312826920)
- [x] Fill out tasks following [discussion](https://github.com/vega/altair/issues/3500)
- [x] Add callout re: `v5.4.0` change [comment](https://github.com/vega/altair/issues/3500#issuecomment-2295148166)
	- Probably makes most sense here   <details><summary><code>parameters.rst</code></summary>  https://github.com/vega/altair/blob/5207768b6e533c0509218376942309d1c7bac22f/doc/user_guide/interactions/parameters.rst?plain=1#L148-L155  </details> 
- [x] `when(..., empty=False)` example in *User Guide* [comment](https://github.com/vega/altair/pull/3544#issuecomment-2329471941)
	- Possibly link to [example](https://altair-viz.github.io/gallery/interactive_bar_select_highlight.html)
	- Similar to the *zorder* link at the end of [conditional-encodings](https://altair-viz.github.io/user_guide/interactions.html#conditional-encodings)
	- Resolved in [`df5555c` (#3544)](https://github.com/vega/altair/pull/3544/commits/df5555cd869c4b4b4b7bf638da08850346110915)
- [x] `parameters.rst`
	- [x] Draft 1
	- [x] Draft 2
- [x] Replace `condition` references with `when` [comment](https://github.com/vega/altair/issues/3500#issuecomment-2295148166)
	- [x] User Guide 
	    - [x]  Interactions
	    	- [x] `bindings_widgets.rst`
	    	- [x] `expressions.rst`
	    	- [x] `jupyter_chart.rst`
	    - [x] Marks
	    	- [x] `geoshape.rst`
	    	- [x] `line.rst`
	    	- [x] `text.rst`
	    - [x] Transform
	    	- [x] `filter.rst`
	    	- [x] `pivot.rst`
	    - [x] `compound_charts.rst`
	    - [x] `times_and_dates.rst`
   - [x] Case Studies
     - [x] `exploring-weather.rst`
   - [x] `README.md`
   - [x] Gallery
   		- [x] `examples_methods_syntax/*.py`
   		- [x] `examples_arguments_syntax/*.py`

dangotbanned commented:
@joelostblom since your mention of this PR in https://github.com/vega/altair/discussions/3548#discussioncomment-10536857, I've circled back to this.

If you have the time, would you be able to double-check the tasks I've added in the [description](https://github.com/vega/altair/pull/3544#issue-2471983178) please?

Going to start working through them now

joelostblom commented:
Looks great! The location for the callout seems suitable to me. The only thing I would add is that I think we should include an example in the user guide section for the use of empty inside the conditional, as opposed to in the parameter definition (https://github.com/vega/altair/pull/3490).

dangotbanned commented:
> Looks great! The location for the callout seems suitable to me. The only thing I would add is that I think we should include an example in the user guide section for the use of empty inside the conditional, as opposed to in the parameter definition (#3490).

Perfect thanks @joelostblom, I've added that to the list.

Managed to get through a good chunk today, I'm expecting `parameters.rst` will require the most work since it calls out `alt.condition` directly

joelostblom reviewed (commented):
I had a peak and it is looking great so far! One comment regarding syntax, curious to hear what you think.

dangotbanned commented on doc/case_studies/exploring-weather.rst:
```diff
@@ -226,33 +226,33 @@ of the selection (for more information on selections, see
-        brush
-    ).properties(
-        width=600
+    points = (
+        alt.Chart()
+        .mark_point()
+        .encode(
+            alt.X("temp_max:Q").title("Maximum Daily Temperature (C)"),
+            alt.Y("temp_range:Q").title("Daily Temperature Range (C)"),
+            color=color,
+            size=alt.Size("precipitation:Q").scale(range=[1, 200]),
+        )
+        .transform_calculate("temp_range", "datum.temp_max - datum.temp_min")
+        .properties(width=600, height=400)
+        .add_params(brush)
+    )
+    bars = (
+        alt.Chart()
+        .mark_bar()
+        .encode(x="count()", y="weather:N", color=alt.Color("weather:N").scale(scale))
+        .transform_calculate("temp_range", "datum.temp_max - datum.temp_min")
+        .transform_filter(brush)
+        .properties(width=600)
```
Happy to discuss in another issue, I shared a similar opinion to that of @binste in https://github.com/vega/altair/pull/3485#issuecomment-2240025423

Specific to this PR - I won't oppose any formatting changes you make, but I would prefer to leave that until review.
I'm also pasting in from `ruff`, so these style choices are inherited from https://github.com/vega/altair/blob/main/pyproject.toml

joelostblom replied:
Sounds good, I opened a discussion in https://github.com/vega/altair/discussions/3570

dangotbanned replied:
@joelostblom I'm "resolving" this since I don't want it to be the focus for review

Please know that I'm not dismissing formatting changes.

As I said in https://github.com/vega/altair/pull/3544#discussion_r1745587004, feel free to make those directly! 👍

dangotbanned commented on tests/examples_methods_syntax/lasagna_plot.py:
```diff
@@ -23,8 +23,8 @@
             labelOverlap=False,
-            labelColor=color_condition,
-            tickColor=color_condition,
+            labelColor=color_condition, # type: ignore[call-overload]
+            tickColor=color_condition, # pyright: ignore[reportArgumentType]
```
Leaving these ignore comments in for reference.

Ideally fix the issue that requires these for a `TypedDict` **before** merging this PR.

dangotbanned replied:
The solve will be using `Map` instead of `dict` in the property setter annotations

dangotbanned replied:
Will be fixed by #3573

dangotbanned commented on tests/examples_arguments_syntax/lasagna_plot.py:
```diff
@@ -23,8 +23,8 @@
             labelOverlap=False,
-            labelColor=color_condition,
-            tickColor=color_condition,
+            labelColor=color_condition, # type: ignore[arg-type]
+            tickColor=color_condition, # type: ignore[arg-type]
```
Related to https://github.com/vega/altair/pull/3544#discussion_r1745791727

Required different ignore comments, since these aren't for property method overload(s)

dangotbanned replied:
Will be fixed by #3573

joelostblom commented:
@dangotbanned I noticed that you tagged me for review here, but also that the PR is still marked as draft, so I just want to double check whether you want a review now or you are waiting to finish something up first.

dangotbanned commented:
> @dangotbanned I noticed that you tagged me for review here, but also that the PR is still marked as draft, so I just want to double check whether you want a review now or you are waiting to finish something up first.

Thanks for checking in @joelostblom!

I was hoping to get feedback focused on the [**Status**](https://github.com/vega/altair/pull/3544#issue-2471983178) section of the description. 

Do you think I should move this out of draft - to make it clearer that I'm interested in a review?

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -146,42 +146,61 @@ we created a selection parameter using ``brush = alt.selection_interval()``,
+The ``"Origin:N"`` specifies how to color the points which fall within the selection,
+and the ``alt.value('lightgray')`` specifies that the outside points should be given a constant color value.
+
+The ``when-then-otherwise`` syntax was directly inspired by `polars.when`_, 
+and is similar to an ``if-else`` statement written in Python::
+
+    # alt.when(brush)
+    if brush:
+        # .then("Origin:N")
+        color = "Origin:N"
+    else:
+        # .otherwise(alt.value("lightgray"))
+        color = alt.value("lightgray")
+    
```
This illustration makes the explanation really easy to understand!

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -296,7 +318,7 @@ selection:
         return alt.Chart(cars).mark_rect().encode(
             x="Cylinders:O",
             y="Origin:N",
-            color=alt.condition(selector, 'count()', alt.value('lightgray'))
+            color=alt.when(brush).then("count()").otherwise(alt.value("lightgray")),
         ).properties(
```
None of the charts in this section rendered for me locally for some reason. Just a note to double check in the final review

dangotbanned replied:
Interesting, I'm just rebuilding the docs now to see if I can spot the issue

dangotbanned replied:
Fixed in [fix: Use `selector` argument in `make_example`](https://github.com/vega/altair/pull/3544/commits/fb42da7de606ec03b8d58773479374492f6e93be)

joelostblom reviewed (commented):
Thanks for your patience @dangotbanned and sorry for not getting here earlier. I had a quick read in the docs and think this is looking great already!

I think it would be instructor to include the following examples in a subsection at the end of "Conditions and Filters":

- Simple conditions may be expressed without defining a default (maybe with a simpler example than `invalid=None`)
- Chain calls to express precise queries
- Reuse defined when variable

Whereas the following seem more like "another way of doing something we already have shown" and I would leave them out of the User Guide (maybe a link a the bottom of the new subsection mentioning there are even more example in the `when` API reference?)

- Predicates passed as positional arguments will be reduced with `&` (since this is just an alternative way of achieving something)
- Using keyword-argument `constraints` can simplify compositions (since this is only for `AND` and `==`)

dangotbanned commented:
> Thanks for your patience @dangotbanned and sorry for not getting here earlier. I had a quick read in the docs and think this is looking great already!

No worries @joelostblom, thanks for taking the time to review

> I think it would be instructor to include the following examples in a subsection at the end of "Conditions and Filters":
> 
> * Simple conditions may be expressed without defining a default (maybe with a simpler example than `invalid=None`)
> * Chain calls to express precise queries
> * Reuse defined when variable

Sounds good to me.
Just to be sure I've understood correctly, which of these locations did you mean?
1. End of [Conditional Encodings](https://github.com/dangotbanned/altair/blob/fb42da7de606ec03b8d58773479374492f6e93be/doc/user_guide/interactions/parameters.rst?plain=1#L259)
	- The end of the `alt.when` section of *Conditions & Filters*
2. End of [Conditions & Filters](https://github.com/dangotbanned/altair/blob/fb42da7de606ec03b8d58773479374492f6e93be/doc/user_guide/interactions/parameters.rst?plain=1#L300)
	- The actual end of *Conditions & Filters*

> Whereas the following seem more like "another way of doing something we already have shown" and I would leave them out of the User Guide (maybe a link a the bottom of the new subsection mentioning there are even more example in the `when` API reference?)
> 
> * Predicates passed as positional arguments will be reduced with `&` (since this is just an alternative way of achieving something)
> * Using keyword-argument `constraints` can simplify compositions (since this is only for `AND` and `==`)

That makes sense, I definitely want to avoid adding too much here - especially if as you've said there is already a documented way to do it

dangotbanned commented on tests/examples_methods_syntax/multiline_tooltip.py:
```diff
@@ -41,15 +41,16 @@
 )
+when_near = alt.when(nearest)
 
 # Draw points on the line, and highlight based on selection
 points = line.mark_point().encode(
-    opacity=alt.condition(nearest, alt.value(1), alt.value(0))
+    opacity=when_near.then(alt.value(1)).otherwise(alt.value(0))
 )
 
 # Draw text labels near the points, and highlight based on selection
 text = line.mark_text(align="left", dx=5, dy=-5).encode(
-    text=alt.condition(nearest, "y:Q", alt.value(" "))
+    text=when_near.then("y:Q").otherwise(alt.value(" "))
```
@joelostblom I decided this would be best left in the gallery in [docs: Add new subsection content](https://github.com/vega/altair/pull/3544/commits/50a520e9314108d572ef8bd6d48f0135151d2c6c)

All of the cases I've reused a (`When|Then`) have been for quite long/detailed examples - which seem excessive to include for just this small bit of functionality

joelostblom commented:
While reviewing https://github.com/vega/altair/pull/3544/commits/50a520e9314108d572ef8bd6d48f0135151d2c6c, I got some ideas for rearranging sections on this page to make it easier to follow. I have most of it done and will finish it off tomorrow to show you my suggestion instead of trying to explain it.

dangotbanned commented:
> While reviewing [50a520e](https://github.com/vega/altair/commit/50a520e9314108d572ef8bd6d48f0135151d2c6c), I got some ideas for rearranging sections on this page to make it easier to follow. I have most of it done and will finish it off tomorrow to show you my suggestion instead of trying to explain it.

All good with me @joelostblom 

Feel free to move this PR out of draft when you're ready.
I was waiting until either I came up with a heading for *Untitled More When* or you took a look

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -428,41 +508,154 @@ cylinders:
-we can construct a scenario where there are two people who can make an interval
+.. altair-plot::
+
+    # empty=False ensure that no points are selected before a selection is drawn
+    brush = alt.selection_interval(empty=False)
+    brush2 = alt.selection_interval(empty=False)
+
```
Note that this and the next example will not quite work as intended until https://github.com/vega/altair/issues/3598 is fixed, but I don't think that is blocking for this PR.

joelostblom replied:
Good point, I added back the old example and will put this up as a separate PR.

joelostblom reviewed (commented):
@dangotbanned To be able to fully document the behavior of `when`,
I made a few organizational changes in https://github.com/vega/altair/pull/3544/commits/e363abdfb2fb1c16a0afaf6338c268f286771e5c

- Split "Conditions" and "Filter" since these are two
  different concept anyways.
- I also renamed the first heading to illustratet that 
  the three big concepts are "Parameter", "Conditions", and "Filters".
- Named the main new "when" section "Understanding when()"
  and included an example of conditional branches in pseudocode early on
  and a runnable example later.
- I also added an additional subheading for "Linking across charts"
- Included mention of the use of `empty=False` inside `when` together
  with a short pseudocode example.
- Creating a new section at the end about combining parameters,
  which contains both the previous section on parameter composition,
  but also two new subsection on multiple conditions and conditional
  when branches.
- I expanded on the parameter composition examples as well.

Please have a look and let me know what you think
I moved the PR out of draft and will go through your other examples today or tomorrow.
I noticed a few unrelated doc changes that I will include in a follow up PR after we merge this.

dangotbanned reviewed (commented):
I'm liking the reorganization @joelostblom!

I picked up a few `rst` corrections that need to be made, and some questions/suggestions on content.

Overall I think we're very close, really appreciate your help with this PR 🤝

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -428,41 +508,154 @@ cylinders:
+        empty=False,
     )
 
+    either_or_but_not_both = (alex | morgan) & ~(alex & morgan)
```
I've gone ahead and updated it to `exlusive_or`. Initially i was thinking it might be confusing for people from a non CS background, but I don't think the one I picked was much clearer anyways =)

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -146,42 +149,104 @@ we created a selection parameter using ``brush = alt.selection_interval()``,
+        # .otherwise(alt.value("lightgray"))
+        color = alt.value("lightgray")
+
+As a convenience,
+simple conditions may be expressed without defining a default:
```
Good point! I like your wording in the docstring and switched to use the same one here.

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -146,42 +149,104 @@ we created a selection parameter using ``brush = alt.selection_interval()``,
+    )
+    points
+
+Multiple conditional branches (``elif`` in Python)
```
Agree, I updated to your first suggestion

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -146,42 +149,104 @@ we created a selection parameter using ``brush = alt.selection_interval()``,
+For now, you can use this pseudocode
+as an example of what the syntax would look like::
+
+    alt.when(brush1)
+    .then(alt.value("goldenrod"))
+    .when(brush2)
+    .then(alt.value("crimson"))
+    .otherwise(alt.value("grey"))
```
Yes that was my intention, but looking at this again I agree it is not really necessary; a link to that section is sufficient so I removed this non-pseudocode example

joelostblom commented on doc/user_guide/interactions/parameters.rst:
```diff
@@ -328,6 +399,15 @@ empty selection contains none of the points:
    interval_x = alt.selection_interval(encodings=['x'], empty=False)
    make_example(interval_x)
 
+The ``empty=False`` argument could instead be set inside :func:`when`,
+so that you don't need to redefine the same selection, e.g.::
```
That was my understanding too and I have tried to update the wording to be more clear

joelostblom reviewed (commented):
Thanks for your helpful feedback on the sections I edited @dangotbanned and sorry it has taken me so long to get back to this. Work has gotten the best of me lately =/ I believe I addressed all your comments so we should be good to go for that section. In terms of reviewing the rest of the examples you have changed, I looked through some of the ones where there is only a few lines changed and they seem good (but I will go through more thoroughly, hopefully tomorrow). As for the larger changes where the formatting of the code has also changed, do you want to change to the syntax we are using currently in the docs or push forward the discussion in https://github.com/vega/altair/discussions/3570 for potentially changing the default syntax? I would like to avoid that we have a mix of the two, but don't feel too strongly for which one we pick. I forget if you had joined the call when we talked briefly about this and there was a couple of points brought up regarding keeping the old syntax because it might be more similar to how new people write who are not familiar with autoformatters and thus easier for them, but nothing is set in stone I believe.

dangotbanned commented:
> Thanks for your helpful feedback on the sections I edited @dangotbanned and sorry it has taken me so long to get back to this. Work has gotten the best of me lately =/ I believe I addressed all your comments so we should be good to go for that section. In terms of reviewing the rest of the examples you have changed, I looked through some of the ones where there is only a few lines changed and they seem good (but I will go through more thoroughly, hopefully tomorrow).

No problem @joelostblom, thanks for updating.

All looking good to me now, I'd hit approve now if I wasn't the author 😄

dangotbanned commented:
> As for the larger changes where the formatting of the code has also changed, do you want to change to the syntax we are using currently in the docs or push forward the discussion in #3570 for potentially changing the default syntax? I would like to avoid that we have a mix of the two, but don't feel too strongly for which one we pick. I forget if you had joined the call when we talked briefly about this and there was a couple of points brought up regarding keeping the old syntax because it might be more similar to how new people write who are not familiar with autoformatters and thus easier for them, but nothing is set in stone I believe.

Ah yeah it would seem I missed that part of the call, apologies @joelostblom.

If there were any changes in positions/consensus then updating #3570 would be helpful for visibility.

Regarding this PR (and reiterating https://github.com/vega/altair/pull/3544#discussion_r1746899291) I'm happy for you to make any formatting changes that you think would help readability 👍

dangotbanned commented:
@joelostblom I wanted to check in with you, as I'm hoping to merge this PR soon.

The following are complicated somewhat by this remaining open:

- #3657
	- Really needs this PR merged first
- #3643
	- I've been holding off on suggesting to move forward on this
	- Likely needs to update multiple places in the user guide
- #3654
	- Not a huge issue, but documenting will touch similar areas

---

I'm thinking pragmatically here, if the only *possibly* outstanding concerns were regarding formatting - would you be okay if we tackled that in a follow up?

I'm really happy with these updates we put together and wish to build on them further in the other issues/PRs I linked

dangotbanned commented:
@mattijn if you're available

I'm planning to merge this tomorrow and follow it up with another PR mentioned in https://github.com/vega/altair/pull/3664#discussion_r1824434520

The short rundown is that it will be adding support all the classes like this:
- `core.Field...Predicate`
- `core.Logical...Predicate`

when used in both of these calls:

- `alt.when(*more_predicates)`
- `alt.Chart.transform_filter(*more_predicates)`

---

I can delay this if anyone would have an issue - but as I mentioned in https://github.com/vega/altair/pull/3544#issuecomment-2445055293 - I'd really like to close this PR off if possible

mattijn commented:
I haven't had a look to this PR before, but just had a quick glance. I understand the title to promote the when-then-otherwise in the user guide, but (unintentionally?) you _ruffed_ many examples into another style that we did not like to promote in the examples.

mattijn commented:
During NumFOCUS summit @binste and I had come up with a config object based on YAPF that comes close to the style of the Altair examples how one would write it manually. 

One can test that config style as such:
```python
import yapf
from yapf.yapflib.yapf_api import FormatCode

# Define the code you want to format
code = """
def example_function(x,y):
 return (x+y)* (x-y)
"""

# Define YAPF configuration
style_config = {
    'based_on_style': 'pep8',
    'COLUMN_LIMIT': 88,
    'SPLIT_BEFORE_EXPRESSION_AFTER_OPENING_PAREN': True,
    'SPLIT_BEFORE_FIRST_ARGUMENT': True,
    'ALLOW_SPLIT_BEFORE_DEFAULT_OR_NAMED_ASSIGNS': True,
    'NO_SPACES_AROUND_SELECTED_BINARY_OPERATORS': False,
    'DEDENT_CLOSING_BRACKETS': True,
    'COALESCE_BRACKETS': True,
    'SPLIT_BEFORE_DOT': True,
    'SPLIT_BEFORE_CLOSING_BRACKET': True,
}

# Format the code using YAPF
formatted_code, _ = FormatCode(code, style_config=style_config)

# Display formatted code
print(formatted_code)
```

I've to mention that this style is not endorsed by @binste, he just prefer to ruff it all! 😄

joelostblom commented:
@dangotbanned Sorry for the delay here. I wanted to go in an revert the examples to the old syntax, but I've been buried in work lately =/ I don't want to hold up your future PRs, so I'm ok if this needs to be merged although if there is something that can be done to bring it closer to the old syntax (like what Mattijn is suggesting maybe?) then that might save time long term instead of trying to change these back in a follow up PR.

As I mentioned in https://github.com/vega/altair/discussions/3570 I do see the advantages of adhering to a syntax that is supported by Ruff, and if our current syntax is making it hard to update the docs effectively, then that's another argument in favor of switching.

dangotbanned reviewed (commented):
Thanks for getting back to me @joelostblom (
https://github.com/vega/altair/pull/3544#issuecomment-2451020864), and @mattijn for your input (https://github.com/vega/altair/pull/3544#issuecomment-2450907421)

I've marked up every instance I spotted that might be considered a change in syntax.

The good news is that @joelostblom's [original comment](https://github.com/vega/altair/pull/3544#discussion_r1744256302) that picked up on the change meant that **most** of this is limited to the `.rst` examples.
So despite this PR touching **64** files; only **10** of these may require changes.

---

### yapf
@mattijn I haven't looked into this in much detail yet, but have some thoughts.

Since it has both a [CLI](https://github.com/vega/altair/pull/3630#issue-2564960629) and [module interface](https://github.com/vega/altair/pull/3630#issue-2564960629) - there may be multiple ways we could use it with the [code block](https://github.com/vega/altair/pull/3630#issue-2564960629) stuff I added in #3630.

I called out https://github.com/vega/altair/discussions/3570#discussioncomment-10560859 directly there, as this was the exact use case I had in mind.
If we went down that route, I'd still prefer to keep that separate from this PR.

I do want to [reiterate](https://github.com/vega/altair/discussions/3570#discussioncomment-10572171) that I feel more strongly towards *some* form of autoformatting for examples.
My preference is `ruff`, but within reason, I'm happy for this to be anything that others agree upon

joelostblom commented:
Thank you for going through and reverting the style on all of these examples before merging @dangotbanned !
